L1Trigger/L1CaloTrigger/plugins/Phase2L1CaloPFClusterEmulator.cc

```diff
@@ -0,0 +1,266 @@
+
+ Description: Creates 3x3 PF clusters from GCTintTowers to be sent to correlator. Follows firmware logic, creates 8 clusters per (2+17+2)x(2+4+2).
+
+ Implementation: To be run together with Phase2L1CaloEGammaEmulator.
+
+*/
+//
+// Original Author:  Pallabi Das
+//         Created:  Wed, 21 Sep 2022 14:54:20 GMT
+//
+//
+
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/L1TCalorimeterPhase2/interface/CaloCrystalCluster.h"
+#include "DataFormats/L1TCalorimeterPhase2/interface/CaloTower.h"
+#include "DataFormats/L1TCalorimeterPhase2/interface/CaloPFCluster.h"
+#include "DataFormats/L1Trigger/interface/EGamma.h"
+
+#include <ap_int.h>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <stdio.h>
+#include "Phase2L1CaloPFClusterEmulator.h"
+
+//#include "Phase2L1RCT.h"
+//#include "Phase2L1GCT.h"
+//#include "Phase2L1GCT_algo.h"
+
+
+//
+// class declaration
+//
+
+class Phase2L1CaloPFClusterEmulator : public edm::stream::EDProducer<> {
+public:
+  explicit Phase2L1CaloPFClusterEmulator(const edm::ParameterSet&);
+  ~Phase2L1CaloPFClusterEmulator();
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void beginStream(edm::StreamID) override;
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  void endStream() override;
+
+  // ----------member data ---------------------------
+  edm::EDGetTokenT<l1tp2::CaloTowerCollection> caloTowerToken_;
+};
+
+//
+// constants, enums and typedefs
+//
+
+
+//
+// static data member definitions
+//
+
+//
+// constructors and destructor
+//
+Phase2L1CaloPFClusterEmulator::Phase2L1CaloPFClusterEmulator(const edm::ParameterSet& iConfig) :
+  caloTowerToken_(consumes<l1tp2::CaloTowerCollection>( iConfig.getParameter<edm::InputTag>( "gctFullTowers")))
+{
+  produces<l1tp2::CaloPFClusterCollection>( "GCTPFCluster" ) ;
+}
+
+Phase2L1CaloPFClusterEmulator::~Phase2L1CaloPFClusterEmulator() {
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+void Phase2L1CaloPFClusterEmulator::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace edm;
+  std::unique_ptr<l1tp2::CaloPFClusterCollection> pfclusterCands(new l1tp2::CaloPFClusterCollection);
+
+  edm::Handle<std::vector<l1tp2::CaloTower>> caloTowerCollection;
+  if(!iEvent.getByToken(caloTowerToken_, caloTowerCollection)) edm::LogError("Phase2L1CaloPFClusterEmulator") << "Failed to get towers from caloTowerCollection!" ;
+
+  iEvent.getByToken(caloTowerToken_, caloTowerCollection);
+  float GCTintTowers[34][72];
+  float realEta[34][72];
+  float realPhi[34][72];
```

Comment:
`34` and `72` are repeated a few times here. It might be helpful to define constants if these numbers have a non-obvious meaning